[ServiceWorkers] Create a better algoritm to match urls (#1165858)

flaki opened the pull request:
Preliminary implementation of the new route matching algorithm + tests:
* I have added basic support for the `*` token which matches arbitrary-length substrings (and is basically translated to `(.*?)`)
* at the moment the `:foobar123`-style tokens are currently simple aliases for the `*` token, their role might later be more refined (arbitrary matching rules, or more restricted translation like `(\w+)`
* I have removed backward compatibility with paths specified with RegExp-syntax

There are two disabled tests, until we decide what to do in the situations:
* Placeholder-crowding (`**`, `:foo:bar` etc.)
* Invalid characters/syntax in path string

arcturus commented:
Could you add some documentation to the readme.md to showcase the changes made to the router?

So people will have a user manual to write their paths

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,116 @@ describe('Router instances', function () {
   });
 
+
+  describe('Reworked matching algorithm (#1165858)', function() {
```
Change the name of the suite. Simply put `Matching algorithm`.

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,116 @@ describe('Router instances', function () {
+    describe('match()', function() {
+      var mw = function() {};
+
+      it('should match trivial registrations (*)', function() {
```
I would not call a path with a wildcard a trivial registration, can you reword this test, please?

flaki replied:
"universal" might be a better fit? since it supposed to match any paths.

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,116 @@ describe('Router instances', function () {
+      it('infix placeholders (/foo/*/bar etc.) should match arbitrary path chunks', function() {
+        var mw2 = function () {};
+
+        router.get('/foo/*', mw);
```
This is the same code as before. :stuck_out_tongue_closed_eyes:

flaki replied:
oops! :blush:

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,116 @@ describe('Router instances', function () {
+
+      it('named placeholders (:foo) should match as anonymous placeholders (*)', function() {
+        router.get('/:foo', mw);
+        expect(router.match('get', '/')).to.deep.equal([mw]);
```
Wow. I'm not sure here. I know we said `:foo` was simply an alias for * right now but by using `:foo` you expect something at least. So I'm not sure if this should match, at least, one character. What do you think @arcturus ?

arcturus replied:
mmm, IMO, :foo should match something like .+, so if we match just the route '/' what could be the potential value of a variable :foo, null? 

I think that when you name a variable you are expect it to exists, is a condition, otherwise you will define a different route.

flaki replied:
sure thing, made it so it's like as described above (no empty matches allowed)

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,116 @@ describe('Router instances', function () {
+        expect(router.match('get', '/foo/bar')).to.deep.equal([mw]);
+      });
+
+      xit('placeholder crowding (**) should ??? ', function() {
```
We agreed on making this the same as `*`.

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,116 @@ describe('Router instances', function () {
+        //expect(router.match('get', '/')).to.be.?;
+        //throw? degrade gracefully? treat as single placeholder? other?
+      });
+
```
We must add the throwing test for `:foo:bar`.

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,116 @@ describe('Router instances', function () {
+    });
+
+    describe('_parseSimplePath()', function() {
+      xit('on invalid registrations it should ???', function() {
```
Let's throw.

lodr replied:
Let's throw. @arcturus ?

arcturus replied:
+1, better to throw with a descriptive message so the error is easy to identify

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,116 @@ describe('Router instances', function () {
+      });
+
+      it('should support anonymous placeholders (*)', function() {
+        expect(router._parseSimplePath('/a*b').source).to.equal("\\/a(.*?)b");
```
Try to stress this method a little bit more. For instance, what if I want an asterisk?

Should I use `('/a\*b')`?

flaki replied:
I have put escaping in there, but since the final (replaced) string is interpreted as a Regular Expression, this is a tricky question.

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,116 @@ describe('Router instances', function () {
+        expect(router._parseSimplePath('/a*b').source).to.equal("\\/a(.*?)b");
+      });
+
+      it('should support named placeholders (:foo)', function() {
```
Same here.

flaki replied:
The colon has no special meaning RegExp-wise so this is fine to escape simply with a backslash.

lodr commented on README.md:
```diff
@@ -47,6 +47,33 @@ worker.post('/event/*', analyticsMiddleware);
 
+## Specifying routes
+
+While registering your middlewares, you are not limited just to predefined paths, you have the choice to specify your routes by using *placeholders*. These placeholders are loosely based on [Express' path strings](http://expressjs.com/guide/routing.html#route-paths).
```
I want a second opinion about the use of the name _placeholder_ for this feature. I got the point, I like it but I preffer to hear from @arcturus as well.

lodr commented on README.md:
```diff
@@ -47,6 +47,33 @@ worker.post('/event/*', analyticsMiddleware);
+
+Currently supported placeholders include:
+
+* Anonymous placeholder: `*`  
```
It could be great if you provide a couple of examples for each placeholder.

lodr commented on README.md:
````diff
@@ -47,6 +47,33 @@ worker.post('/event/*', analyticsMiddleware);
+worker.get('/:path'); // will match /foo and /foo/bar, but won't match /
+```
+
+You could use the backslash character to escape special placeholders (and specify literal asterisks and/or colon characters in your code). *Note: in JavaScript string literals you must double the backslash to achieve the intended effect.*
````
I think this is better explained by using exposing the problem first and then providing the solution. It's just an opinion. @arcturus ?

lodr replied:
Wow. Double scaping could make things really, really ugly. What about using `#` or `?` as scaping character? Some other else? I would prefer to write `'/foo/#*'` rather tahn `'/foo/\\*'` I choose them because these characters are not allowed in the path so if the user wants to use them as literals inside the URL, they must use the encoded versions: `%23` and `%3f`.

A second opinion here, please? @arcturus @flaki ?

flaki replied:
I don't think going against the convention (of backslash being an escaping character) is a good idea. Escaping really is over-insurance here, as (as you said so) I wouldn't expect either `*` or `:` to be frequently used in matching string definitions -- I think we can handle a bit of ugliness in the rare cases they *do* need escaped. I don't feel comfortable creating a new escaping grammar just for this purpose, especially as there are other ways around this (like you said, URLEncoding into entities). Also, most experienced JS devs will be well aware of the backslash needing to be duplicated, so this is more of a "just-in-case" comment in the docs.
That said I'm fine with anything, @arcturus any thoughts on this?

lodr replied:
Fair to me. Not blocking here.

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,142 @@ describe('Router instances', function () {
+    describe('match()', function() {
+      var mw = function() {};
+
+      it('should match universal registrations (*)', function() {
```
Let's use the same vocabulary `should match anonymous placeholders`. Does it make sense to you?

flaki replied:
ah yeah, missed this one. thanks!

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,142 @@ describe('Router instances', function () {
+        expect(router.match('get', '/foo/bar')).to.deep.equal([mw]);
+      });
+
+      it('should match all universal registrations (*) in order', function() {
```
Reword to `should match all universal registrations (*) preserving registration order`

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,142 @@ describe('Router instances', function () {
+      it('placeholder crowding for named placeholder (:foo:bar) should throw', function() {
+        expect(function () { router.get('/foo/:bar:baz', mw); }).to.throw(Error, 'Invalid usage of named placeholders');
+      });
+
```
Add a couple of more tests to test _triple_ crowding: `***` and `:foor:bar:foo`.

lodr commented:
We forgot the case of mixed crowdings (`*:foo` or `:foo*`), I would throw in any case. What do you think?

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,142 @@ describe('Router instances', function () {
+        expect(router.match('get', '/pre/suf/inner/fix/')).to.deep.equal([mw, mw2, mw3]);
+      });
+
+      it('named placeholders (:foo) should match similar to anonymous ones', function() {
```
`similar` is a little bit vague. Better: `named placeholders (:foo) should match any sequence of characters except the empty string`.

lodr commented on lib/spec/router.sw-spec.js:
```diff
@@ -206,4 +206,142 @@ describe('Router instances', function () {
+        expect(router.match('get', '/foo/bar')).to.deep.equal([mw]);
+      });
+
+      it('named placeholders (:foo) must match at least one character (no empty matches allowed)', function() {
```
Actually, you can merge this test with the previous one.

lodr commented on lib/router.js:
```diff
@@ -78,4 +84,35 @@ Router.prototype._sanitizeMethod = function(method) {
+  try {
+    // Parsing anonymous placeholders with simple backslash-escapes
+    path = path.replace(/(.|^)\*/g, function(m,escape) {
+      return escape==='\\'?'\\*':(escape+'(.*?)');
```
Add spaces surrounding expressions. It's a little bit hard to read.

lodr commented on lib/router.js:
```diff
@@ -78,4 +84,35 @@ Router.prototype._sanitizeMethod = function(method) {
+
+    // Parsing named placeholders with backslash-escapes
+    path = path.replace(/(.|^)\:([a-zA-Z0-9]+)/g, function(m,escape,tag) {
+      return escape==='\\'?(':'+tag):(escape+'(.+?)');
```
Same here.

lodr commented on lib/router.js:
```diff
@@ -78,4 +84,35 @@ Router.prototype._sanitizeMethod = function(method) {
+  // Try parsing the string and converting special characters into regex
+  try {
+    // Parsing anonymous placeholders with simple backslash-escapes
+    path = path.replace(/(.|^)\*/g, function(m,escape) {
```
What do you want to mean with `/(.|^)/`. Are you sure you're trying to say "[any character or `^` character](https://regex101.com/r/aA2gY3/1)"?

flaki replied:
this denotes ["a character or the start of the string"](http://stackoverflow.com/a/2552440/1178775).

lodr replied:
Yep, I never saw this inside an or expression. :)

flaki commented:
I have fixed the nits, updated the docs for the router, and added tests for triple crowding.
Also, I think all mixed crowdings should throw, so I added tests and updated `router.js` accordingly.
Could you have a look @lodr?
